File: dotnet_bot/src/WealthsimpleTradingBot/Api/MarketDataService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using WealthsimpleTradingBot.Models;

namespace WealthsimpleTradingBot.Api;

public interface IMarketDataService
{
    Task<Security?> SearchSecurityAsync(string symbol);
    Task<decimal> GetQuoteAsync(string symbol);
    Task<string> ResolveSecurityIdAsync(string symbol);
    Task<Dictionary<string, string>> BulkResolveSecuritiesAsync(List<string> symbols);
    Task<Dictionary<string, decimal>> GetBulkQuotesAsync(List<string> symbols);
}

public class MarketDataService : IMarketDataService
{
    private readonly IWealthsimpleClient _client;
    private readonly ConcurrentDictionary<string, Security> _cache = new();
    private readonly ILogger<MarketDataService> _logger;

    public MarketDataService(IWealthsimpleClient client, ILogger<MarketDataService> logger)
    {
        _client = client;
        _logger = logger;
    }

    public async Task<Security?> SearchSecurityAsync(string symbol)
    {
        if (_cache.TryGetValue(symbol, out var cached))
            return cached;

        var query = symbol.Replace(".TO", "");
        var data = await _client.GetAsync<JsonElement>("/securities",
            new Dictionary<string, string> { ["query"] = query });

        if (!data.TryGetProperty("results", out var results))
            return null;

        foreach (var raw in results.EnumerateArray())
        {
            var rawSymbol = GetNestedString(raw, "stock", "symbol");
            if (rawSymbol.Equals(query, StringComparison.OrdinalIgnoreCase) ||
                rawSymbol.Equals(symbol, StringComparison.OrdinalIgnoreCase))
            {
                var sec = ParseSecurity(raw);
                _cache[symbol] = sec;
                return sec;
            }
        }

        // Take first result as fallback
        if (results.GetArrayLength() > 0)
        {
            var sec = ParseSecurity(results[0]);
            _cache[symbol] = sec;
            return sec;
        }

        _logger.LogWarning("Security not found: {Symbol}", symbol);
        return null;
    }

    public async Task<decimal> GetQuoteAsync(string symbol)
    {
        var sec = await SearchSecurityAsync(symbol);
        return sec?.Quote?.Amount ?? 0;
    }

    public async Task<string> ResolveSecurityIdAsync(string symbol)
    {
        var sec = await SearchSecurityAsync(symbol);
        return sec?.Id ?? throw new InvalidOperationException(
            $"Could not resolve security ID for {symbol}");
    }

    public async Task<Dictionary<string, string>> BulkResolveSecuritiesAsync(List<string> symbols)
    {
        var result = new Dictionary<string, string>();
        foreach (var symbol in symbols)
        {
            try
            {
                result[symbol] = await ResolveSecurityIdAsync(symbol);
            }
            catch (Exception ex)
            {
                _logger.LogWarning("Could not resolve {Symbol}: {Error}", symbol, ex.Message);
            }
        }
        _logger.LogInformation("Resolved {Count}/{Total} securities",
            result.Count, symbols.Count);
        return result;
    }

    public async Task<Dictionary<string, decimal>> GetBulkQuotesAsync(List<string> symbols)
    {
        var result = new Dictionary<string, decimal>();
        foreach (var symbol in symbols)
        {
            var price = await GetQuoteAsync(symbol);
            if (price > 0) result[symbol] = price;
        }
        return result;
    }

    private static Security ParseSecurity(JsonElement raw)
    {
        var stock = raw.TryGetProperty("stock", out var s) ? s : raw;

        SpotQuote? quote = null;
        if (raw.TryGetProperty("quote", out var q))
        {
            quote = new SpotQuote(
                Amount: GetDecimal(q, "amount"),
                Ask: q.TryGetProperty("ask", out var ask) ? ask.GetDecimal() : null,
                Bid: q.TryGetProperty("bid", out var bid) ? bid.GetDecimal() : null,
                High: GetDecimal(q, "high"),
                Low: GetDecimal(q, "low"),
                Volume: q.TryGetProperty("volume", out var vol) ? vol.GetInt64() : 0,
                QuoteDate: GetString(q, "quote_date")
            );
        }

        return new Security(
            Id: GetString(raw, "id"),
            Symbol: GetString(stock, "symbol"),
            Name: GetString(stock, "name"),
            Exchange: GetString(stock, "primary_exchange"),
            Currency: GetString(stock, "currency", "CAD"),
            SecurityType: GetString(stock, "security_type"),
            IsBuyable: stock.TryGetProperty("is_buyable", out var ib) && ib.GetBoolean(),
            Quote: quote
        );
    }

    private static string GetString(JsonElement el, string name, string def = "")
        => el.TryGetProperty(name, out var val) ? val.GetString() ?? def : def;

    private static string GetNestedString(JsonElement el, string parent, string child)
    {
        if (el.TryGetProperty(parent, out var p) && p.TryGetProperty(child, out var c))
            return c.GetString() ?? "";
        return "";
    }

    private static decimal GetDecimal(JsonElement el, string name)
        => el.TryGetProperty(name, out var val) ? val.GetDecimal() : 0;
}

File: python_bot/src/market_data_service.py
import logging
from decimal import Decimal
from typing import Optional, Dict, List
from .models import Security, SpotQuote

logger = logging.getLogger(__name__)


class MarketDataService:
    """Wraps Wealthsimple security search and quote endpoints."""

    def __init__(self, client):
        self._client = client
        self._cache: Dict[str, Security] = {}

    def search_security(self, symbol: str) -> Optional[Security]:
        if symbol in self._cache:
            return self._cache[symbol]

        # Strip .TO suffix for WS search
        query = symbol.replace(".TO", "")
        data = self._client.get("/securities", params={"query": query})
        results = data.get("results", [])

        for raw in results:
            raw_symbol = raw.get("stock", {}).get("symbol", "")
            if raw_symbol.upper() == query.upper() or raw_symbol.upper() == symbol.upper():
                sec = self._parse_security(raw)
                self._cache[symbol] = sec
                return sec

        # If exact match not found, take first result
        if results:
            sec = self._parse_security(results[0])
            self._cache[symbol] = sec
            return sec

        logger.warning("Security not found: %s", symbol)
        return None

    def get_security_by_id(self, security_id: str) -> Optional[Security]:
        data = self._client.get(f"/securities/{security_id}")
        return self._parse_security(data) if data else None

    def get_quote(self, symbol: str) -> Decimal:
        sec = self.search_security(symbol)
        if sec and sec.quote:
            return sec.quote.amount
        return Decimal("0")

    def resolve_security_id(self, symbol: str) -> str:
        sec = self.search_security(symbol)
        if sec:
            return sec.id
        raise ValueError(f"Could not resolve security ID for {symbol}")

    def bulk_resolve_securities(self, symbols: List[str]) -> Dict[str, str]:
        result = {}
        for symbol in symbols:
            try:
                result[symbol] = self.resolve_security_id(symbol)
            except ValueError:
                logger.warning("Could not resolve: %s", symbol)
        logger.info("Resolved %d/%d securities", len(result), len(symbols))
        return result

    def get_bulk_quotes(self, symbols: List[str]) -> Dict[str, Decimal]:
        result = {}
        for symbol in symbols:
            price = self.get_quote(symbol)
            if price > 0:
                result[symbol] = price
        return result

    def _parse_security(self, raw: dict) -> Security:
        stock = raw.get("stock", raw)
        quote_data = raw.get("quote", {})
        quote = None
        if quote_data:
            quote = SpotQuote(
                amount=Decimal(str(quote_data.get("amount", 0))),
                ask=Decimal(str(quote_data.get("ask", 0))) if quote_data.get("ask") else None,
                bid=Decimal(str(quote_data.get("bid", 0))) if quote_data.get("bid") else None,
                high=Decimal(str(quote_data.get("high", 0))),
                low=Decimal(str(quote_data.get("low", 0))),
                volume=int(quote_data.get("volume", 0)),
                quote_date=str(quote_data.get("quote_date", "")),
            )

        return Security(
            id=raw.get("id", stock.get("id", "")),
            symbol=stock.get("symbol", ""),
            name=stock.get("name", ""),
            exchange=stock.get("primary_exchange", ""),
            currency=stock.get("currency", "CAD"),
            security_type=stock.get("security_type", ""),
            is_buyable=stock.get("is_buyable", True),
            quote=quote,
        )
